keps/draft-20190516-KUDO-package-manager.md

```diff
@@ -0,0 +1,177 @@
+* Having a solution on how to distribute this structure
+* Having this platform-agnostic so Packages can be distributed across multiple solutions
+* Adapting concepts from other successfull open-source projects in this regard
+
+### Non-Goals
+
+Non-goals are:
+* Re-inventing the wheel
+* Solving problems that are not related to the overall goals
+* Some more that I can't think of yet
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is. The folder structure and what could be interpreted as our v0 of the Registry was originally discussed in PR [#87](https://github.com/kudobuilder/kudo/pull/87) and is a great way to start when thinking of this enhancement.
+
+### User Stories
+
+#### Framework Developer
+
+As a Framework Developer I, ...
+
+* would like to be able to host validation/testing plans and resources as part of the package
+* would like to update my package version
+
+#### Cluster Administrator
+
+As a Cluster Administrator I, ...
+
+* would like to be able to host my custom Framework extensions on a private KUDO Framework repo
+* would like to be able to add additional validation/testing plans and resources
+
+### Implementation Details/Notes/Constraints
+
+We need to make sure that Packages can be easily accessed even if the underlying Storage Backend changed. Therefor, adding a middle layer like a registry server that has multiple Storage Backends as a swappable engine underneath seems like a good pattern. This also would enable users to keep track or restrict access to specific Frameworks.
+
+The idea would be a HTTP Server that can be easily accessed, even in air-gapped environments, depending on the user needs. As Storage Backends there could be a multitude such as `Local`, `Google`, `S3`, `Github`, `Minio`, `Docker` and so on.
+Having as an API a simple HTTP Server that just serves essentially a single yaml file that holds all the information also would enable us as a first iteration to use any HTTP server. This should work as it already with Google Cloud Storage or S3, all it requires is having access to the yaml file and something like the KUDO registry that serves like a middle layer inbetween could be easily added later on.
+
+Some caveats to this could for instance be defining a clean interface that won't break things when a user attempts to access Frameworks and hosted repos. Another caveat is deciding on the right proper structure in how a Package and its higher repo structure will look like. Design decisions here could potentialy impact future implementations of Storage Backends we haven't thought of yet. Another caveat would be identifying the right approach in versioning as this dictates a lot also how the structure will look like. Versioning for the `.tgz` files and Frameworks should be in SemVer.
+
+The Repo structure on your local laptop could look in the short term as follow:
+
+```bash
+/index.yaml
+/kafka
+/kafka/2.2.0
+/kafka/2.2.0/kafka-framework.yaml
```

Comment:
this structure is not true anymore however. I feel like we should remove the package format from this KEP since that is what KEP9 already defined. I like the repository structure though, so 👍 to that

Reply:
I've added it for short term but also then long term to paint a picture how the structure actually will look like. If it changes, this would supersede this KEP or directly change it. In any way I think having it here contributes to more and more formalizing how the structure would look like. Otherwise it could be maybe harder to understand or grasp.

Reply:
See also https://github.com/kudobuilder/frameworks/issues/17